app/v2/htlc/internal/keeper/keeper.go

```diff
@@ -0,0 +1,95 @@
+package keeper
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/irisnet/irishub/app/v1/params"
+	"github.com/irisnet/irishub/app/v2/htlc/internal/types"
+	"github.com/irisnet/irishub/codec"
+	sdk "github.com/irisnet/irishub/types"
+)
+
+type Keeper struct {
+	storeKey sdk.StoreKey
+	cdc      *codec.Codec
+	bk       types.BankKeeper
+
+	// codespace
+	codespace sdk.CodespaceType
+	// params subspace
+	paramSpace params.Subspace
+}
+
+func NewKeeper(cdc *codec.Codec, key sdk.StoreKey, bk types.BankKeeper, codespace sdk.CodespaceType, paramSpace params.Subspace) Keeper {
+	return Keeper{
+		storeKey:   key,
+		cdc:        cdc,
+		bk:         bk,
+		codespace:  codespace,
+		paramSpace: paramSpace.WithTypeTable(types.ParamTypeTable()),
+	}
+}
+
+// Codespace returns the codespace
+func (k Keeper) Codespace() sdk.CodespaceType {
+	return k.codespace
+}
+
+// CreateHTLC creates a HTLC
+func (k Keeper) CreateHTLC(ctx sdk.Context, htlc types.HTLC) (sdk.Tags, sdk.Error) {
+	// get the secret hash lock
+	secretHash := htlc.GetSecretHashLock()
+
+	// check if the secret hash lock already exists
+	if k.HasSecretHashLock(ctx, secretHash) {
+		return nil, types.ErrSecretHashLockAlreadyExists(types.DefaultCodespace, "the secret hash lock already exists")
+	}
+
+	// lock the specified tokens
+	// TODO
+
+	// set the htlc
+	k.SetHTLC(ctx, htlc)
+
+	createTags := sdk.NewTags(
+		types.TagSender, []byte(htlc.Sender),
+		types.TagReceiver, []byte(htlc.Sender),
+		types.TagReceiverOnOtherChain, []byte(htlc.ReceiverOnOtherChain),
+		types.TagOutAmount, []byte(htlc.OutAmount.String()),
+		types.TagInAmount, sdk.Uint64ToBigEndian(htlc.InAmount),
+		types.TagSecretHashLock, []byte(hex.EncodeToString(htlc.GetSecretHashLock())),
+		types.TagTimestamp, sdk.Uint64ToBigEndian(htlc.Timestamp),
+		types.TagExpireHeight, sdk.Uint64ToBigEndian(htlc.ExpireHeight),
```

Comment:
Too many tags, I think only `sender`,`receiver`, `secretHashLock` is necessary.

Reply:
I think, if want to generate HTLC on another chain automatically with tools, these tags are necessary.